### PR TITLE
add RandIntLayer

### PR DESCRIPTION
**Pull request by vieting**
Add `RandIntLayer` which accepts a shape based on `DimensionTag`s or ints (only for static dims) and creates random ints using `tf.random.uniform`.

### CODE REVIEW

File: tests/test_TFNetworkLayer.py [hunk 1]
```diff
@@ -1568,6 +1568,37 @@ def test_ConvLayer_get_valid_out_dim():
...
+        "data": {"dim": n_out}}
+    })
+    net = TFNetwork(config=config, train_flag=True)
+    n_batch = 3
+    max_seq_len = 10
+    feed = make_feed_dict(net.extern_data.data.values(), n_batch=n_batch, n_time=max_seq_len, same_time=True)
+    size_placeholder = net.extern_data.data["data"].size_placeholder[0]
+    input_len = feed[size_placeholder]
+    sz = (
+      DimensionTag(description="batch", kind=DimensionTag.Types.Batch),
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:1587]:
```suggestion
      DimensionTag(kind=DimensionTag.Types.Batch),
```

File: tests/test_TFNetworkLayer.py [hunk 2]
```diff
@@ -1568,6 +1568,37 @@ def test_ConvLayer_get_valid_out_dim():
...
+      DimensionTag(description="time", kind=DimensionTag.Types.Time, dyn_size=size_placeholder),
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:1588]:
```suggestion
      net.extern_data.data["data"].get_size_dim_tag(0),
```

File: tests/test_TFNetworkLayer.py [hunk 3]
```diff
@@ -1568,6 +1568,37 @@ def test_ConvLayer_get_valid_out_dim():
...
+      DimensionTag(description="feature", kind=DimensionTag.Types.Feature, dimension=5),
```
Comment (albertz) [inline tests/test_TFNetworkLayer.py:1589]:
```suggestion
      5,
```

File: returnn/tf/layers/basic.py [hunk 4]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+
+  @classmethod
+  def get_out_data_from_opts(cls, name, shape, maxval, minval=0, dtype="int32", **kwargs):
+    """
+    :param str name:
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :param int maxval: upper bound (exclusive) on range of random values
+    :param int minval: lower bound (inclusive) on range of random values
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param kwargs:
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1713]:
```suggestion
```

File: returnn/tf/layers/basic.py [hunk 5]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+          shape_parsed.append(s.dimension)
+          feature_dim_axis = ax
+        elif s.dimension is None:
+          shape_parsed.append(None)
+          time_dim_axis = ax
+        else:
+          assert False
+      else:
+        raise NotImplementedError
+    shape_parsed = tuple(shape_parsed)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1739]: You have duplicate code here. Put that into a (class) helper function.

File: returnn/tf/layers/basic.py [hunk 6]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+
+    if batch_dim_axes:
+      assert len(batch_dim_axes) == 1
+      self.output.size_placeholder = {i: size_placeholder for i, size_placeholder in enumerate(dyn_axes_sizes)}
+    else:
+      self.output.size_placeholder = {}
+
+    self.output.placeholder = tf.random.uniform(shape_parsed, minval=minval, maxval=maxval, dtype=dtype, seed=seed)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1693]: In case `seed` is None, you should use sth like `seed=self.network.random.randint(2 ** 31)`. See other related code.

File: returnn/tf/layers/basic.py [hunk 7]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+
+  # noinspection PyUnusedLocal
+  def __init__(self, shape, maxval, minval=0, dtype="int32", seed=None, **kwargs):
+    """
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :param int maxval: upper bound (exclusive) on range of random values
+    :param int minval: lower bound (inclusive) on range of random values
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param int|None seed: random seed
+    :param kwargs:
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1660]:
```suggestion
```

File: returnn/tf/layers/basic.py (hunk 5 above)
Comment (albertz) [reply returnn/tf/layers/basic.py:1739]: Ah sorry, I see now it's not exactly the same.
Then leave it as it is. A helper function would just complicate it more than it helps.

File: returnn/tf/layers/basic.py [hunk 8]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+        elif s.dimension is None:
+          assert s.dyn_size is not None
+          shape_parsed.append(tf.reduce_max(s.dyn_size))
+          dyn_axes_sizes.append(tf.fill(tf.shape(s.dyn_size), shape_parsed[-1]))
+        else:
+          assert False
+      else:
+        raise NotImplementedError
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1684]:
```suggestion
        raise TypeError("%s: invalid dim %s" % (self, type(s)))
```

File: returnn/tf/layers/basic.py [hunk 9]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+          batch_dim_axes.append(ax)
+          shape_parsed.append(self.get_batch_dim())
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+        elif s.dimension is None:
+          assert s.dyn_size is not None
+          shape_parsed.append(tf.reduce_max(s.dyn_size))
+          dyn_axes_sizes.append(tf.fill(tf.shape(s.dyn_size), shape_parsed[-1]))
+        else:
+          assert False
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1682]:
```suggestion
          raise Exception("%s: invalid dim tag %s" % (self, s))
```

File: returnn/tf/layers/basic.py [hunk 10]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+        if s.kind == DimensionTag.Types.Batch:
+          batch_dim_axis = ax
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+          feature_dim_axis = ax
+        elif s.dimension is None:
+          shape_parsed.append(None)
+          time_dim_axis = ax
+        else:
+          assert False
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1736]:
```suggestion
        else:
          assert s.dimension is None
          shape_parsed.append(None)
          time_dim_axis = ax
```

File: returnn/tf/layers/basic.py [hunk 11]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+
+    shape_parsed = []
+    batch_dim_axis = None
+    time_dim_axis = None
+    feature_dim_axis = None
+    for ax, s in enumerate(shape):
+      if isinstance(s, int):
+        shape_parsed.append(s)
+        feature_dim_axis = ax
+      elif isinstance(s, DimensionTag):
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1726]:
```suggestion
      else:
        assert isinstance(s, DimensionTag)
```

File: returnn/tf/layers/basic.py [hunk 12]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+          feature_dim_axis = ax
+        elif s.dimension is None:
+          shape_parsed.append(None)
+          time_dim_axis = ax
+        else:
+          assert False
+      else:
+        raise NotImplementedError
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1738]:
```suggestion
```
(And make the `else:` branch above, as suggested.)

File: returnn/tf/layers/basic.py [hunk 13]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param kwargs:
+    :rtype: Data
+    """
+    from returnn.tf.util.data import DimensionTag, NotSpecified
+
+    shape_parsed = []
+    batch_dim_axis = None
+    time_dim_axis = None
+    feature_dim_axis = None
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1721]: Do not set `feature_dim_axis` nor `time_dim_axis`. Leave the default logic of `Data`.

Comment (albertz) [inline returnn/tf/layers/basic.py:1721]:
```suggestion
```

File: returnn/tf/layers/basic.py [hunk 14]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+      if batch_dim_axis is None or (isinstance(batch_dim_axis, int) and batch_dim_axis > feature_dim_axis):
+        dim = shape_parsed[feature_dim_axis]
+      else:
+        dim = shape_parsed[feature_dim_axis - 1]
+    else:
+      dim = NotSpecified
+
+    return Data(
+      name="%s_output" % name, shape=shape_parsed, dim=dim, dtype=dtype,
+      batch_dim_axis=batch_dim_axis, time_dim_axis=time_dim_axis, feature_dim_axis=feature_dim_axis)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1751]:
```suggestion
    return Data(
      name="%s_output" % name, shape=shape_parsed, dtype=dtype, batch_dim_axis=batch_dim_axis)
```

File: returnn/tf/layers/basic.py [hunk 15]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+    shape_parsed = tuple(shape_parsed)
+
+    if feature_dim_axis is not None:
+      if batch_dim_axis is None or (isinstance(batch_dim_axis, int) and batch_dim_axis > feature_dim_axis):
+        dim = shape_parsed[feature_dim_axis]
+      else:
+        dim = shape_parsed[feature_dim_axis - 1]
+    else:
+      dim = NotSpecified
+
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1748]:
```suggestion
```
Just let the default logic of `Data` handle this.

File: returnn/tf/layers/basic.py [hunk 16]
```diff
@@ -1645,6 +1645,112 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
     return out
 
 
+class RandIntLayer(LayerBase):
+  """Generates random numbers using ``tf.random.uniform``"""
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1649]:
```suggestion
  """
  Generates random numbers using ``tf.random.uniform``
  """
```
(Consistent style.)

File: returnn/tf/layers/basic.py [hunk 17]
```diff
@@ -1645,6 +1645,98 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+      elif isinstance(s, DimensionTag):
+        if s.kind == DimensionTag.Types.Batch:
+          batch_dim_axes.append(ax)
+          shape_parsed.append(self.get_batch_dim())
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+        elif s.dimension is None:
+          assert s.dyn_size is not None
+          shape_parsed.append(tf.reduce_max(s.dyn_size))
+          dyn_axes_sizes.append(tf.fill(tf.shape(s.dyn_size), shape_parsed[-1]))
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1683]: Why?
`s.dyn_size` is what you should use and put into `size_placeholder`.
```suggestion
          dyn_axes_sizes.append(s.dyn_size)
```

Comment (albertz) [reply returnn/tf/layers/basic.py:1683]: But actually, you also need to remember the axis.
```suggestion
          dyn_axes_sizes[ax] = s.dyn_size
```

File: returnn/tf/layers/basic.py [hunk 18]
```diff
@@ -1645,6 +1645,98 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+    :param int|None seed: random seed
+    """
+    from returnn.tf.util.data import DimensionTag
+    super(RandIntLayer, self).__init__(**kwargs)
+
+    seed = seed if seed is not None else self.network.random.randint(2 ** 31)
+
+    shape_parsed = []
+    batch_dim_axes = []
+    dyn_axes_sizes = []
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1670]:
```suggestion
    dyn_axes_sizes = {}
```

File: returnn/tf/layers/basic.py [hunk 19]
```diff
@@ -1645,6 +1645,98 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+          dyn_axes_sizes.append(tf.fill(tf.shape(s.dyn_size), shape_parsed[-1]))
+        else:
+          raise Exception("%s: invalid dim tag %s" % (self, s))
+      else:
+        raise TypeError("%s: invalid dim %s" % (self, type(s)))
+    shape_parsed = tuple(shape_parsed)
+
+    if batch_dim_axes:
+      assert len(batch_dim_axes) == 1
+      self.output.size_placeholder = {i: size_placeholder for i, size_placeholder in enumerate(dyn_axes_sizes)}
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1692]:
```suggestion
      self.output.size_placeholder = {self.output.get_batch_axis_excluding_batch(i): size for i, size in dyn_axes_sizes.items()}
```

File: returnn/tf/layers/basic.py [hunk 20]
```diff
@@ -1645,6 +1645,98 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+    else:
+      self.output.size_placeholder = {}
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1694]: The `size_placeholder` actually should not be set here, but already in `get_out_data_from_opts` below.

Comment (vieting) [reply returnn/tf/layers/basic.py:1694]: Is there a particular reason for that here? Most layers set the `size_placeholder` in the init.

File: returnn/tf/layers/basic.py (hunk 17 above)
Comment (vieting) [reply returnn/tf/layers/basic.py]: I figured for the random numbers the sequence length does not matter, but I agree it's probably better this way.

File: returnn/tf/layers/basic.py (hunk 20 above)
Comment (vieting) [reply returnn/tf/layers/basic.py]: I did it now anyway.

Comment (albertz) [reply returnn/tf/layers/basic.py]: No. Actually almost all layers set it in `get_out_data_from_opts`. Only those layers who create a *new* dyn size, they do it in `__init__` (because we don't want to modify the TF graph in `get_out_data_from_opts`, i.e. add new ops in there). But in your case, you don't add anything new. You just reuse existing dyn sizes / existing `DimensionTag`s.

Setting them in `get_out_data_from_opts` is important, because currently they carry the `DimensionTag` information for the dynamic axes, and we need this information for various functions.

**Comment by vieting** [comment]
I moved the parsing to a class helper function now. The comments regarding `size_placeholder` were addressed, it is also set in `get_out_data_from_opts` now.

File: returnn/tf/layers/basic.py (hunk 20 above)
Comment (vieting) [reply returnn/tf/layers/basic.py]: Alright, thanks!

File: returnn/tf/layers/basic.py [hunk 21]
```diff
@@ -1645,6 +1645,113 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+      elif isinstance(s, DimensionTag):
+        if s.kind == DimensionTag.Types.Batch:
+          assert batch_dim_axis is None
+          batch_dim_axis = ax
+          shape_parsed.append(s)
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+        elif s.dimension is None:
+          assert s.dyn_size is not None
+          shape_parsed.append(tf.reduce_max(s.dyn_size))
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1711]: There should be no tensor operations in `get_out_data_from_opts`.
I'm not sure it is helpful to have this helper function for both `get_out_data_from_opts` and `__init__`. At least this makes it quite complicated.

File: returnn/tf/layers/basic.py [hunk 22]
```diff
@@ -1645,6 +1645,113 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+  @classmethod
+  def _parse_shape(cls, shape):
+    """
+    Parse the desired shape for further usage in init and get_out_data_from_opts.
+    The returned shape includes batch dim as DimensionTag and dynamic axes as tensors.
+    Additionally, return dict of dynamic axes and their sizes.
+
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :return: parsed shape including batch dim and sizes of the dynamic axes
+    :rtype: tuple[tuple|dict]
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1692]: The type is wrong.
You mean this?
```suggestion
    :rtype: (tuple[int|DimensionTag], dict[int,tf.Tensor])
```

File: returnn/tf/layers/basic.py [hunk 23]
```diff
@@ -1645,6 +1645,113 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
...
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param int|None seed: random seed
+    """
+    from returnn.tf.util.data import DimensionTag
+    super(RandIntLayer, self).__init__(**kwargs)
+
+    seed = seed if seed is not None else self.network.random.randint(2 ** 31)
+
+    shape_parsed, _ = self._parse_shape(shape)
+    shape_parsed = tuple(s if not isinstance(s, DimensionTag) else self.get_batch_dim() for s in shape_parsed)
```
Comment (albertz) [inline returnn/tf/layers/basic.py:1669]: This is very unintuitive.
You could maybe pass `batch_dim` to the `_parse_shape` function. But don't do such post edit here.

**Comment by albertz** [comment]
> I moved the parsing to a class helper function now. 

I think that actually makes the code more complicated than it needs to be. See also my other comments.

**Comment by vieting** [comment]
I undid most of the last changes, including the class helper function, and addressed the previous comments in a more limited way.

**Review by albertz — APPROVED** [review]
